t/docs/lib/python/urlvalidator.py

```diff
@@ -0,0 +1,184 @@
+            return validate_url(url, full, verify=False)
+        return CONNECTION_ERROR
+    except requests.exceptions.HTTPError:
+        return PAGE_NOT_FOUND
+    else:
+        code = int(response.status_code)
+        if code in [301, 302, 303, 308]:
+            #errors.append((url, file_, URL_REDIRECT))
+            pass
+        if code == 405 and not full:
+            # If we get a "method not allowed" try using a get request (slow).
+            return validate_url(url, full=True)
+        elif code < 200 or code >= 300:
+            return HTTP_ERROR.format(code)
+    return None
+
+def report_invalid_urls(int_errors, ext_errors, frg_errors, malformed):
+    """Print out list of errors."""
+    errors_by_file = {}
+    for url, file_, error in int_errors + ext_errors + frg_errors + malformed:
+        if file_ not in errors_by_file:
+            errors_by_file[file_] = []
+        errors_by_file[file_].append((url, error,))
+
+    if errors_by_file:
+        print >> sys.stderr, 'Errors found in the following files:'
+    for file_, errors in errors_by_file.iteritems():
+        print >> sys.stderr
+        print >> sys.stderr, '\t', file_
+        for url, error in errors:
+            print >> sys.stderr, '\t\t"{url}"  -  "{error}"'.format(
+                url=url, error=error)
+
+def main(root):
+    """Extract and check all URLs on the site located at root."""
+    external = {}
+    internal = []
+    fragments = []
+    malformed = []
+    for file_ in get_files_by_extension(root, 'html'):
+        with open(file_, 'r') as html_file:
+            matches = URL_PATTERN.findall(html_file.read())
+            for url in matches:
+                if url.startswith('mailto:'):
+                    # We aren't validating email addresses.
+                    continue
+                if url.startswith('javascript:'):
+                    # We arent validating javascript functions.
+                    continue
+                if url.startswith('#') and len(url) > 1:
+                    # URL fragment (on current page).
+                    fragments.append((url[1:], file_, file_))
+                elif url.startswith('www.'):
+                    malformed.append((url, file_, INVALID_URL))
+                elif url.startswith('http'):
+                    if url not in external:
+                        external[url] = file_
+                else:
+                    internal.append((url, file_,))
+
+    int_errors, new_frags = validate_internal(internal)
+    ext_errors = validate_external(external)
+    frg_errors = validate_fragments(new_frags + fragments)
+
+    report_invalid_urls(int_errors, ext_errors, frg_errors, malformed)
+
+    if int_errors or ext_errors or frg_errors or malformed:
+        sys.exit(1)
+    sys.exit(0)
+
+if __name__ == '__main__':
+    if len(sys.argv) != 2:
+        print ('Usage: Please supply root directory to validate the contents '
+               'of.')
```

Comment:
Might as well check we have the correct number of arguments and provide some form of CLI documentation to future developers.